nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -180,11 +180,18 @@ impl DataStore {
         let updated = diesel::update(dsl::instance)
             .filter(dsl::time_deleted.is_null())
             .filter(dsl::id.eq(*instance_id))
-            .filter(dsl::state_generation.lt(new_runtime.gen))
+            // Runtime state updates are allowed if either:
+            // - the active Propolis ID will not change, the state generation
+            //   increased, and the Propolis generation will not change, or
```

Comment:
Nice, it's impressive you pulled this off while keeping a single "runtime update" function. It's nice this logic is centralized.